File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js


function solution(maps) {
    
    const dy = [-1, 0, 0, 1];
    const dx = [0, -1, 1, 0];
    
    const n = maps.length-1;
    const m = maps[0].length-1;
    const visited = Array.from({length: n+1}, () => Array(m+1).fill(false));
    
    const queue = [];
    queue.push([0, 0, 1]);
    visited[0][0] = true;
    
    while(queue.length) {
        
        const [y, x, blocks] = queue.shift();
        if (y === n && x === m) return blocks;
        
        for (let i = 0; i < 4; i++) {
            const ty = y + dy[i];
            const tx = x + dx[i];
            
            if (ty < 0 || tx < 0 || ty > n || tx > m || !maps[ty][tx] || visited[ty][tx]) continue;
            
            visited[ty][tx] = true;
            queue.push([ty, tx, blocks+1]);
            
        }
        
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.Arrays;
import java.util.Deque;
import java.util.ArrayDeque;

class Solution {
    public static int r, c;
    
    public static boolean isAvailable(int nr, int nc) {
        if (nr < 0 || nc < 0 || nr >= r || nc >= c) return false;
        return true;
    }
    
    public int solution(int[][] map) {
        
        r = map.length;
        c = map[0].length;
        
        int[][] dist = new int[r][c];
        
        for (int i = 0; i < r; i++) {
            Arrays.fill(dist[i], 0);
        }
        
        int[] dr = {-1, 0, 0, 1};
        int[] dc = {0, -1, 1, 0};
        
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{0, 0});
        dist[0][0] = 1;
        
        while (!queue.isEmpty()) {
            int[] point = queue.poll();

            int crntR = point[0];
            int crntC = point[1];
            int crntDist = dist[crntR][crntC];
            
            for (int i = 0; i < 4; i++) {
                int newR = crntR + dr[i];
                int newC = crntC + dc[i];
                if (isAvailable(newR, newC) && map[newR][newC] == 1) {
                    if (dist[newR][newC] > crntDist + 1 || dist[newR][newC] == 0) {
	                    queue.add(new int[]{newR, newC});
                        dist[newR][newC] = crntDist + 1;
                    }
                }
            }
        }
        
        if (dist[r-1][c-1] == 0) return -1;
        
        return dist[r-1][c-1];
    }
}